zebra-state/src/on_disk.rs

```diff
@@ -295,10 +355,66 @@ impl_from! {
     SerializationError,
     std::io::Error,
     sled::Error,
+    broadcast::RecvError,
 }
 
 impl Into<BoxError> for Error {
     fn into(self) -> BoxError {
         BoxError::from(self.0)
     }
 }
+
+/// A map like interface ontop of sled with a future based interface for
+/// retrieving values that aren't available yet.
+#[allow(dead_code)]
+struct SledMap<K, V> {
+    tree: Tree,
+    pending: HashMap<K, broadcast::Sender<V>>,
+}
+
+#[allow(clippy::manual_async_fn)]
+#[allow(dead_code, unused_variables)]
+impl<K, V> SledMap<K, V>
+where
+    K: AsRef<[u8]> + Eq + Hash,
+    V: ZcashDeserialize + ZcashSerialize + Clone,
+{
+    fn new(tree: Tree, pending: HashMap<K, broadcast::Sender<V>>) -> Self {
+        Self { tree, pending }
+    }
+
+    fn insert(&mut self, key: K, value: V) -> impl Future<Output = Option<V>> {
+        async { todo!() }
+    }
+
+    fn get(&mut self, k: K) -> impl Future<Output = Option<V>> {
+        // if contains outpoint, return outpoint, if not, check if we have a
+        // channel for it, if so subscribe and return the receiver, if not,
+        // create a channel, insert, and return the receiver
+        let val = self
+            .tree
+            .get(&k)
+            .unwrap()
+            .map(|bytes| bytes.zcash_deserialize_into().unwrap())
+            .map(Either::Left)
+            .unwrap_or_else(|| {
+                Either::Right(
+                    self.pending
+                        .get(&k)
+                        .map(|sender| sender.subscribe())
+                        .unwrap_or_else(|| {
+                            let (tx, rx) = broadcast::channel(1);
+                            self.pending.insert(k, tx);
+                            rx
+                        }),
+                )
+            });
+
+        async {
+            match val {
+                Either::Left(val) => Some(val),
+                Either::Right(mut rx) => Some(rx.recv().await.unwrap()),
+            }
```

Comment:
I don't think we can return the UTXO until we have the previous block, because any block before the current block can spend the UTXO.